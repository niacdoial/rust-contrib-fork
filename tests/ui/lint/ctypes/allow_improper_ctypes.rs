
#![deny(improper_ctypes, improper_c_fn_definitions, improper_ctype_definitions, improper_c_callbacks)]

//@ aux-build: outer_crate_types.rs
//@ compile-flags:--extern outer_crate_types
extern crate outer_crate_types as outer;

// ////////////////////////////////////////////////////////
// first, the same bank of types as in the extern crate

#[repr(C)]
struct SafeStruct (i32);

#[repr(C)]
struct UnsafeStruct (String);
//~^ ERROR: repr(C) type uses type `String`

#[repr(C)]
#[allow(improper_ctype_definitions)]
struct AllowedUnsafeStruct (String);

// refs are only unsafe if the value comes from the other side of the FFI boundary
// due to the non-null assumption
// (technically there are also assumptions about non-dandling, alignment, aliasing, lifetimes, etc...)
// the lint is not raised here, but will be if used in the wrong place
#[repr(C)]
struct UnsafeFromForeignStruct<'a> (&'a u32);

#[repr(C)]
#[allow(improper_ctype_definitions)]
struct AllowedUnsafeFromForeignStruct<'a> (&'a u32);


type SafeFnPtr = extern "C" fn(i32)->i32;

type UnsafeFnPtr = extern "C" fn((i32, i32))->i32;
//~^ ERROR: `extern` callback uses type `(i32, i32)`

#[allow(improper_c_callbacks)]
type AllowedUnsafeFnPtr = extern "C" fn(&[i32])->i32;

type UnsafeRustCalleeFnPtr = extern "C" fn(i32)->&'static i32;
//~^ ERROR: `extern` callback uses type `&i32`

#[allow(improper_c_callbacks)]
type AllowedUnsafeRustCalleeFnPtr = extern "C" fn(i32)->&'static i32;

// for now, let's not lint on the nonzero assumption,
// because we don't know if the callback is rust-callee-foreign-caller or the other way around
// and having to cast around function signatures to get function pointers would be an awful experience
// let's assume that the unsafety in this fnptr will be pointed out indirectly by a lint elsewhere
type UnsafeForeignCalleeFnPtr = extern "C" fn(&i32);

#[allow(improper_c_callbacks)]
type AllowedUnsafeForeignCalleeFnPtr = extern "C" fn(&i32);


// ////////////////////////////////////////////////////////
// then, some functions that use them

static INT: u32 = 42;

#[allow(improper_c_fn_definitions)]
extern "C" fn fn1a(e: &String) -> &str {&*e}
extern "C" fn fn1u(e: &String) -> &str {&*e}
//~^ ERROR: `extern` fn uses type `&str`
//~^^ ERROR: `extern` fn uses type `&String`

#[allow(improper_c_fn_definitions)]
extern "C" fn fn2a(e: UnsafeStruct) {}
extern "C" fn fn2u(e: UnsafeStruct) {}
//~^ ERROR: `extern` fn uses type `UnsafeStruct`
#[allow(improper_c_fn_definitions)]
extern "C" fn fn2oa(e: outer::UnsafeStruct) {}
extern "C" fn fn2ou(e: outer::UnsafeStruct) {}
//~^ ERROR: `extern` fn uses type `outer::UnsafeStruct`

#[allow(improper_c_fn_definitions)]
extern "C" fn fn3a(e: AllowedUnsafeStruct) {}
extern "C" fn fn3u(e: AllowedUnsafeStruct) {}  // FIXME: ...ideally the lint should not trigger here
//~^ ERROR: `extern` fn uses type `AllowedUnsafeStruct`
#[allow(improper_c_fn_definitions)]
extern "C" fn fn3oa(e: outer::AllowedUnsafeStruct) {}
extern "C" fn fn3ou(e: outer::AllowedUnsafeStruct) {}  // FIXME: ...ideally the lint should not trigger here
//~^ ERROR: `extern` fn uses type `outer::AllowedUnsafeStruct`

#[allow(improper_c_fn_definitions)]
extern "C" fn fn4a(e: UnsafeFromForeignStruct) {}
extern "C" fn fn4u(e: UnsafeFromForeignStruct) {}
//~^ ERROR: `extern` fn uses type `UnsafeFromForeignStruct<'_>`
#[allow(improper_c_fn_definitions)]
extern "C" fn fn4oa(e: outer::UnsafeFromForeignStruct) {}
extern "C" fn fn4ou(e: outer::UnsafeFromForeignStruct) {}
//~^ ERROR: `extern` fn uses type `outer::UnsafeFromForeignStruct<'_>`

#[allow(improper_c_fn_definitions)]
extern "C" fn fn5a() -> UnsafeFromForeignStruct<'static> { UnsafeFromForeignStruct(&INT)}
extern "C" fn fn5u() -> UnsafeFromForeignStruct<'static> { UnsafeFromForeignStruct(&INT)}
#[allow(improper_c_fn_definitions)]
extern "C" fn fn5oa() -> outer::UnsafeFromForeignStruct<'static> { outer::UnsafeFromForeignStruct(&INT)}
extern "C" fn fn5ou() -> outer::UnsafeFromForeignStruct<'static> { outer::UnsafeFromForeignStruct(&INT)}

#[allow(improper_c_fn_definitions)]
extern "C" fn fn6a() -> AllowedUnsafeFromForeignStruct<'static> { AllowedUnsafeFromForeignStruct(&INT)}
extern "C" fn fn6u() -> AllowedUnsafeFromForeignStruct<'static> { AllowedUnsafeFromForeignStruct(&INT)}
#[allow(improper_c_fn_definitions)]
extern "C" fn fn6oa() -> outer::AllowedUnsafeFromForeignStruct<'static> { outer::AllowedUnsafeFromForeignStruct(&INT)}
extern "C" fn fn6ou() -> outer::AllowedUnsafeFromForeignStruct<'static> { outer::AllowedUnsafeFromForeignStruct(&INT)}

fn main(){}
