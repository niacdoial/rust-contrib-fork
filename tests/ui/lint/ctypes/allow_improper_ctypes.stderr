error: repr(C) type uses type `String`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:15:1
   |
LL | struct UnsafeStruct (String);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`UnsafeStruct`) is FFI-unsafe due to a `String` field
note: the type is defined here
  --> $DIR/allow_improper_ctypes.rs:15:1
   |
LL | struct UnsafeStruct (String);
   | ^^^^^^^^^^^^^^^^^^^
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout
note: the lint level is defined here
  --> $DIR/allow_improper_ctypes.rs:2:53
   |
LL | #![deny(improper_ctypes, improper_c_fn_definitions, improper_ctype_definitions, improper_c_callbacks)]
   |                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `extern` callback uses type `(i32, i32)`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:36:20
   |
LL | type UnsafeFnPtr = extern "C" fn((i32, i32))->i32;
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: the function pointer to `extern "C" fn((i32, i32)) -> i32` is FFI-unsafe due to `(i32, i32)`
   = help: consider using a struct instead
   = note: tuples have unspecified layout
note: the lint level is defined here
  --> $DIR/allow_improper_ctypes.rs:2:81
   |
LL | #![deny(improper_ctypes, improper_c_fn_definitions, improper_ctype_definitions, improper_c_callbacks)]
   |                                                                                 ^^^^^^^^^^^^^^^^^^^^

error: `extern` callback uses type `&i32`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:42:30
   |
LL | type UnsafeRustCalleeFnPtr = extern "C" fn(i32)->&'static i32;
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: the function pointer to `extern "C" fn(i32) -> &i32` is FFI-unsafe due to `&i32`
   = help: consider using a raw pointer, or wrapping `&i32` in an `Option<_>`
   = note: boxes and references are assumed to be valid (non-null, non-dangling, aligned) pointers,
           which cannot be garanteed if their values are produced by non-rust code

error: `extern` fn uses type `&String`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:65:23
   |
LL | extern "C" fn fn1u(e: &String) -> &str {&*e}
   |                       ^^^^^^^ not FFI-safe
   |
   = help: consider using a raw pointer, or wrapping `&String` in an `Option<_>`
   = note: boxes and references are assumed to be valid (non-null, non-dangling, aligned) pointers,
           which cannot be garanteed if their values are produced by non-rust code
note: the lint level is defined here
  --> $DIR/allow_improper_ctypes.rs:2:26
   |
LL | #![deny(improper_ctypes, improper_c_fn_definitions, improper_ctype_definitions, improper_c_callbacks)]
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^

error: `extern` fn uses type `&str`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:65:35
   |
LL | extern "C" fn fn1u(e: &String) -> &str {&*e}
   |                                   ^^^^ not FFI-safe
   |
   = help: consider using `*const u8` and a length instead
   = note: this reference to an unsized type contains metadata, which makes it incompatible with a C pointer

error: `extern` fn uses type `UnsafeStruct`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:71:23
   |
LL | extern "C" fn fn2u(e: UnsafeStruct) {}
   |                       ^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`UnsafeStruct`) is FFI-unsafe due to a `String` field
note: the type is defined here
  --> $DIR/allow_improper_ctypes.rs:15:1
   |
LL | struct UnsafeStruct (String);
   | ^^^^^^^^^^^^^^^^^^^
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout

error: `extern` fn uses type `outer::UnsafeStruct`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:75:24
   |
LL | extern "C" fn fn2ou(e: outer::UnsafeStruct) {}
   |                        ^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`outer::UnsafeStruct`) is FFI-unsafe due to a `String` field
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout

error: `extern` fn uses type `AllowedUnsafeStruct`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:80:23
   |
LL | extern "C" fn fn3u(e: AllowedUnsafeStruct) {}  // FIXME: ...ideally the lint should not trigger here
   |                       ^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`AllowedUnsafeStruct`) is FFI-unsafe due to a `String` field
note: the type is defined here
  --> $DIR/allow_improper_ctypes.rs:20:1
   |
LL | struct AllowedUnsafeStruct (String);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout

error: `extern` fn uses type `outer::AllowedUnsafeStruct`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:84:24
   |
LL | extern "C" fn fn3ou(e: outer::AllowedUnsafeStruct) {}  // FIXME: ...ideally the lint should not trigger here
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`outer::AllowedUnsafeStruct`) is FFI-unsafe due to a `String` field
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout

error: `extern` fn uses type `UnsafeFromForeignStruct<'_>`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:89:23
   |
LL | extern "C" fn fn4u(e: UnsafeFromForeignStruct) {}
   |                       ^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`UnsafeFromForeignStruct<'_>`) is FFI-unsafe due to a `&u32` field
note: the type is defined here
  --> $DIR/allow_improper_ctypes.rs:27:1
   |
LL | struct UnsafeFromForeignStruct<'a> (&'a u32);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: consider using a raw pointer, or wrapping `&u32` in an `Option<_>`
   = note: boxes and references are assumed to be valid (non-null, non-dangling, aligned) pointers,
           which cannot be garanteed if their values are produced by non-rust code

error: `extern` fn uses type `outer::UnsafeFromForeignStruct<'_>`, which is not FFI-safe
  --> $DIR/allow_improper_ctypes.rs:93:24
   |
LL | extern "C" fn fn4ou(e: outer::UnsafeFromForeignStruct) {}
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: this struct/enum/union (`outer::UnsafeFromForeignStruct<'_>`) is FFI-unsafe due to a `&u32` field
   = help: consider using a raw pointer, or wrapping `&u32` in an `Option<_>`
   = note: boxes and references are assumed to be valid (non-null, non-dangling, aligned) pointers,
           which cannot be garanteed if their values are produced by non-rust code

error: aborting due to 11 previous errors

